weldx/geometry.py

```diff
@@ -2377,6 +2410,102 @@ def spatial_data(
+            raise NotImplementedError
+
+        shape_raster_data = shape_raster_data.swapaxes(1, 2)
+
+        num_profiles = shape_raster_data.shape[0]
+        num_profile_points = shape_raster_data.shape[1]
+
+        points = shape_raster_data.reshape(
+            (num_profiles * num_profile_points, 3)
+        ).tolist()
+
+        tri_base = []
+        for i in range(num_profile_points):
+            idx_0 = i
+            idx_1 = (i + 1) % num_profile_points
+            idx_2 = idx_0 + num_profile_points
+            idx_3 = idx_1 + num_profile_points
+
+            if cw_ordering:
+                tri_base += [[idx_0, idx_2, idx_1], [idx_1, idx_2, idx_3]]
+            else:
+                tri_base += [[idx_0, idx_1, idx_2], [idx_1, idx_3, idx_2]]
+
+        tri_base = np.array(tri_base, dtype=int)
+        triangles = np.array(
+            [
+                tri_base + i * num_profile_points + offset
+                for i in range(num_profiles - 1)
+            ],
+            dtype=int,
+        )
+
+        triangles = triangles.reshape(
+            (tri_base.shape[0] * (num_profiles - 1), 3)
+        ).tolist()
+
+        # closes front and back faces
+        tri_cw = []
+        tri_ccw = []
+        i_0 = 0
+        i_1 = 0
+
+        while i_0 + i_1 < num_profile_points - 2:
+            if i_1 == i_0:
+                p_0 = i_0 + offset
+                p_1 = p_0 + 1
+                p_2 = num_profile_points + offset - i_1 - 1
+                i_0 += 1
+            else:
+                p_0 = i_0 + offset
+                p_1 = num_profile_points + offset - i_1 - 2
+                p_2 = p_1 + 1
+                i_1 += 1
+            tri_cw += [[p_0, p_1, p_2]]
+            tri_ccw += [[p_0, p_2, p_1]]
+
+        if cw_ordering:
+            front = tri_cw
+            back = tri_ccw
+        else:
+            front = tri_ccw
+            back = tri_cw
+        triangles += front
+        triangles += (
+            np.array(back, int) + (num_profiles - 1) * num_profile_points
+        ).tolist()
+
+        return points, triangles
+
+    def write_to_file(self, file_name: str, profile_raster_width, trace_raster_width):
+        for s in self._trace.segments:
+            if isinstance(s, RadialHorizontalTraceSegment):
+                raise NotImplementedError
+
+        raster_data = self._rasterize_constant_profile(
+            profile_raster_width=profile_raster_width,
+            trace_raster_width=trace_raster_width,
+            stack=False,
+        )
+        points = []
+        triangles = []
+        for i, shape_raster_data in enumerate(raster_data):
+            shape_points, shape_triangle = self._shape_points_and_triangles(
+                shape_raster_data,
+                len(points),
+                self.profile.shapes[i].is_polygon_winding_order_cw(),
+            )
+            points += shape_points
+            triangles += shape_triangle
```

Comment:
can we simply move this part to the end of `Geometry.spatial_data()` ?

The way I understand it now we get a different triangulation when calling `write_to_file()` on a `Geometry` directly or on the `SpatialData` (via `Geometry.rasterize().write_to_file()` )

I think getting a single (correct) triangulation when calling `Geometry.rasterize()` is more intuitive than different triangulations. In that case we would not need the `Geometry.write_to_file()` function but could use the existing `SpatialData.write_to_file()` function

I think the intu way would be

Reply:
Have mostly implemented this now. Just have to replace the branch for single parts. Afterwards, I can remove the old triangulation function (maybe, see below). I just realized that the implementation should also work for variable profiles if the number of points for different profiles stays the same. However, I have to check the rasterization output for variable profiles. I am actually not sure what it returns because the different profiles can have a different number of points and as far as I know, this shouldn't play well with `numpy.array` which is the return type for constant profiles. I came across this when I removed the `VariableProfile` exception, which can't be used the same way in the `SpatialData` as in the `Geometry`.

Apart from that, I will come up with a triangulation version for `VariableProfiles` (already have an idea) or we keep the old one for them.